[quant][graphmode] InsertObservers for shared class types

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #30553 [quant][graphmode] `insert_quant_dequant` pass support shared class types
* #30552 [quant][graphmode][refactor] Add registerQParams function
* #30551 [quant][graphmode] Insert GetAttr for quantization parameters instead of Constant
* #30550 [quant][graphmode] Make `InsertQuantDeQuantHelper` global
* #30549 [quant][graphmode] Rename QuantizeHelper to InsertQuantDeQuantHelper
* **#30548 [quant][graphmode] InsertObservers for shared class types**

Summary:
ClassTypes can be shared among different module instances, but previously we assumed
they would be unique, this PR enables the insert_observers pass to work with shared class types
Test Plan:
python test/test_jit.py
python test/test_quantization.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D18802465](https://our.internmc.facebook.com/intern/diff/D18802465)

jerryzh168 committed:
[quant][graphmode] InsertObservers for shared class types

Summary:
ClassTypes can be shared among different module instances, but previously we assumed
they would be unique, this PR enables the insert_observers pass to work with shared class types
Test Plan:
python test/test_jit.py
python test/test_quantization.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] InsertObservers for shared class types"


Summary:
ClassTypes can be shared among different module instances, but previously we assumed
they would be unique, this PR enables the insert_observers pass to work with shared class types
Test Plan:
python test/test_jit.py
python test/test_quantization.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] InsertObservers for shared class types"


Summary:
ClassTypes can be shared among different module instances, but previously we assumed
they would be unique, this PR enables the insert_observers pass to work with shared class types
Test Plan:
python test/test_jit.py
python test/test_quantization.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] InsertObservers for shared class types"


Summary:
ClassTypes can be shared among different module instances, but previously we assumed
they would be unique, this PR enables the insert_observers pass to work with shared class types
Test Plan:
python test/test_jit.py
python test/test_quantization.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] InsertObservers for shared class types"


Summary:
ClassTypes can be shared among different module instances, but previously we assumed
they would be unique, this PR enables the insert_observers pass to work with shared class types
Test Plan:
python test/test_jit.py
python test/test_quantization.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] InsertObservers for shared class types"


Summary:
ClassTypes can be shared among different module instances, but previously we assumed
they would be unique, this PR enables the insert_observers pass to work with shared class types
Test Plan:
python test/test_jit.py
python test/test_quantization.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] InsertObservers for shared class types"


Summary:
ClassTypes can be shared among different module instances, but previously we assumed
they would be unique, this PR enables the insert_observers pass to work with shared class types
Test Plan:
python test/test_jit.py
python test/test_quantization.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] InsertObservers for shared class types"


Summary:
ClassTypes can be shared among different module instances, but previously we assumed
they would be unique, this PR enables the insert_observers pass to work with shared class types
Test Plan:
python test/test_jit.py
python test/test_quantization.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

ZolotukhinM commented on test/test_jit.py:
```diff
@@ -1151,6 +1152,28 @@ def forward(self, x):
+            'Expected to have 3 observer submodules'
+        assert len([x for x, _ in m.conv2._modules._c.items()
+                    if x.startswith('_observer_')]) == 3, \
+            'Expected to have 3 observer submodules'
```
Could we somehow check that the conv modules have the same type after quantization? Also, just to verify: the test does fail without this change, right?

jerryzh168 replied:
yeah this fails before the change. I didn't find a way check the types are the same though

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -421,6 +424,17 @@ void InsertObserversHelper::insertObservers(
+      module._ivalue()->setAttr(name, observer.clone_instance()._ivalue());
+    }
+    return;
+  }
```
This condition should probably be moved to the beginning of the function. There is no need to repeat this work for every invoked method.

jerryzh168 replied:
what do you mean? 
We didn't repeat this part for every invoked method, this is outside of the loop of calling insertObservers for invoked methods.
this is intentionally put after we call insertObservers for all invoked methods.

ZolotukhinM reviewed (approved):
Looks good!

jerryzh168 committed:
Update on "[quant][graphmode] InsertObservers for shared class types"


Summary:
ClassTypes can be shared among different module instances, but previously we assumed
they would be unique, this PR enables the insert_observers pass to work with shared class types
Test Plan:
python test/test_jit.py
python test/test_quantization.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -178,6 +178,8 @@ class InsertObserversHelper {
   // Set of observer forward call nodes
   std::unordered_set<Node*> observer_nodes_;
+  // Map from type of the module to the vector of observer modules we added to the class
+  std::unordered_map<Graph*, std::vector<std::tuple<std::string, script::Module>>> graph_observer_map_;
```
oh it should be graph, I forgot to change the comment

jerryzh168 replied:
when I dump the module I can find:
   conv1 = <__torch__.torch.nn.modules.conv.Conv2d object at 0x560b745835a0>
    conv2 = <__torch__.torch.nn.modules.conv.Conv2d object at 0x560b745c6490>

jerryzh168 committed:
Update on "[quant][graphmode] InsertObservers for shared class types"


Summary:
ClassTypes can be shared among different module instances, but previously we assumed
they would be unique, this PR enables the insert_observers pass to work with shared class types
Test Plan:
python test/test_jit.py
python test/test_quantization.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]
